### PR TITLE
drivers: mfd: Add Infineon HPPASS MFD driver

### PR DESCRIPTION
**Pull request by jsbatch**
Part 1 of a series of PRs enabling the High Performance Programmable Analog Sub-system on the Infineon PSC3M5 family of MCUs.  This PR adds the top level MFD Driver all other HPPASS drivers are built on.  

HPPASS hosts SAR ADC and Comparator/Slope Generator (CSG) devices which are coordinated by an Autonomous Controller (AC).  The AC is managed with a hardware State Transition Table that can be configured through the MFD driver.  The MFD driver owns the HPPASS init, AC management, trigger I/O routing, and shared HPPASS interrupt handling.  

The driver allows for two operating modes of the AC:
- **Basic** (default): driver synthesizes a minimal STT from the enabled child set to allow basic operation.
- **Advanced**: STT built from an `ac-states` phandle list, one entry per node, for applications that need a custom and more advanced AC program.

### CODE REVIEW

File: dts/bindings/mfd/infineon,hppass-analog.yaml [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+  Infineon HPPASS (High Performance Programmable Analog Sub-System)
+  Multi-Function Device.
+
+  The HPPASS subsystem contains multiple analog peripherals (SAR ADC, CSG)
+  coordinated by an Autonomous Controller (AC) state machine.  This MFD
+  device manages the shared resources: subsystem initialisation, AC lifecycle,
+  STT management, and the combined MCPASS interrupt (AC state + error
+  conditions).
+
+  Individual peripheral drivers (SAR ADC, CSG) are child nodes of this device
```
Comment (JarmouniA) [inline dts/bindings/mfd/infineon,hppass-analog.yaml]:
```suggestion
  Individual peripherals (SAR ADC, CSG) are represented as child nodes of this device
```

**Comment by jsbatch** [comment]
Retrying CI

File: drivers/mfd/Kconfig.infineon_hppass [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 Infineon Technologies AG,
```
Comment (teburd) [inline drivers/mfd/Kconfig.infineon_hppass]: <text>... All rights reserved.</text>

**Review by JarmouniA — COMMENTED** [review]
New driver needs in-tree testing (or a build test at the minimum).

File: dts/bindings/mfd/infineon,hppass-analog.yaml (hunk 1 above)
Comment (jsbatch) [reply dts/bindings/mfd/infineon,hppass-analog.yaml]: This driver implementation is part of a larger analog subsystem and implements some of the core functionality, however I don't know that it would be terribly helpful to test it in isolation.  As part of this development, I've got some more PRs in the works that will use this driver, including an update to the HPPASS ADC driver and an implementation of the CSG module within the HPPASS.  

Can we defer testing until those ADC and/or CSG features are available?

Comment (jsbatch) [reply dts/bindings/mfd/infineon,hppass-analog.yaml]: Updated infineon,hppass-analog.yaml per your suggestion.

File: drivers/mfd/mfd_infineon_hppass_regs.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ *   - Composed constants for register values whose individual fields
+ *     are documented in the TRM but whose @e composition is required
+ *     for normal operation (e.g. AREF_CTRL init value).
+ *   - TRM section citations next to each helper.
+ *
+ * References
+ * ----------
+ *
+ *   - @e PSoC @e Control @e C3 @e Architecture @e TRM, 002-39348 Rev. *C
+ *   - @e PSoC @e Control @e C3 @e Registers @e TRM,    002-39445 Rev. *B
```
Comment (ifx-rmcs) [inline drivers/mfd/mfd_infineon_hppass_regs.h]: HPPASS is available on other devices too. I think you should remove these references

**Review by JarmouniA — DISMISSED** [review]

File: drivers/mfd/mfd_infineon_hppass.c [hunk 4]
```diff
@@ -0,0 +1,1084 @@
...
+ * or an affiliate of Infineon Technologies AG.</text>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief MFD parent for the HPPASS analog subsystem.
+ *
+ * Owns block-wide HPPASS state: subsystem init, the Autonomous Controller
+ * (AC) lifecycle and STT, trigger I/O routing, and the combined MCPASS
```
Comment (ifx-rmcs) [inline drivers/mfd/mfd_infineon_hppass.c]: Replaced MCPASS with HPPASS

File: drivers/mfd/mfd_infineon_hppass.c [hunk 5]
```diff
@@ -0,0 +1,1084 @@
...
+/*
+ * Latching level trigger.  Stays asserted until cleared via
+ * ifx_hppass_ac_clear_fw_trigger_level(); use this for cooperative stops
+ * where the AC must observe the trigger no matter where it is in the STT.
+ */
+void ifx_hppass_ac_set_fw_trigger_level(const struct device *dev, uint8_t mask)
+{
+	const struct ifx_hppass_mfd_config *config = dev->config;
+
+	HPPASS_INFRA_FW_TR_LEVEL(config->base) |= mask;
```
Comment (ifx-rmcs) [inline drivers/mfd/mfd_infineon_hppass.c]: Could use sys_set_bits()

File: drivers/mfd/mfd_infineon_hppass.c [hunk 6]
```diff
@@ -0,0 +1,1084 @@
...
+} /* ifx_hppass_ac_set_fw_trigger_level() */
+
+/* Clear an asserted FW level trigger bit. */
+void ifx_hppass_ac_clear_fw_trigger_level(const struct device *dev, uint8_t mask)
+{
+	const struct ifx_hppass_mfd_config *config = dev->config;
+
+	HPPASS_INFRA_FW_TR_LEVEL(config->base) &= ~(uint32_t)mask;
```
Comment (ifx-rmcs) [inline drivers/mfd/mfd_infineon_hppass.c]: Could use sys_clear_bits()

File: drivers/mfd/mfd_infineon_hppass.c [hunk 7]
```diff
@@ -0,0 +1,1084 @@
...
+		}                                                                                  \
+	}
+
+static int ifx_hppass_late_start(void)
+{
+	DT_INST_FOREACH_STATUS_OKAY(IFX_HPPASS_LATE_START)
+	return 0;
+}
+
+SYS_INIT(ifx_hppass_late_start, APPLICATION, 0);
```
Comment (ifx-rmcs) [inline drivers/mfd/mfd_infineon_hppass.c]: I suppose 0 means it's the highest priority and would set up the MFD before other dependent HPPASS blocks. Is it ok to use 0?

Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass.c]: Good point, the priority shouldn't be hard coded.  I've added a KConfig option for the priority.  0 means this is the highest priority APPLICATION level initialization routine, so it will do the late_start function after the kernel level init functions.

File: drivers/mfd/mfd_infineon_hppass.c (hunk 6 above)
Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass.c]: See above.

File: drivers/mfd/mfd_infineon_hppass.c (hunk 5 above)
Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass.c]: Sure, makes sense.  To keep the entire driver consistent in how it writes registers I've updated all the references to use sys_* calls.

File: drivers/mfd/mfd_infineon_hppass.c (hunk 4 above)
Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass.c]: Fixed

File: drivers/mfd/mfd_infineon_hppass_regs.h (hunk 3 above)
Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass_regs.h]: I get what you mean about HPPASS being available on other devices, but I still think it's valuable to have at least some reference for where the register offsets and IP behavior were derived from.  I've expanded the comments a bit to mention that the references are based off the PSOC Control C3 Mainline device manuals.

**Review by ifx-rmcs — DISMISSED** [review]
Sonarcube seems to have 2 warnings. You might want check those. The rest of the changes look good.

File: drivers/mfd/mfd_infineon_hppass.c [hunk 8]
```diff
@@ -0,0 +1,1088 @@
...
+	struct trim_entry {
+		volatile uint32_t *addr;
+		uint32_t value;
+	};
+	struct trims {
+		uint32_t num_entries;
+		struct trim_entry entry[IFX_HPPASS_TRIMS_ENTRIES_NUM];
+	};
+
+	const struct trims *infra_trims = (const struct trims *)&SFLASH_SAR_INFRA_TRIM_TABLE;
```
Comment (teburd) [inline drivers/mfd/mfd_infineon_hppass.c:253]: Since this is referencing memory elsewhere, this could be a flex array member instead of a fixed sized type if you'd like. When might the num_entries *not* be the defined entry count would be an added question here? It seems like we *know* its 7 entries, we also have the num_entries member to introspect.

File: drivers/mfd/mfd_infineon_hppass.c [hunk 9]
```diff
@@ -0,0 +1,1088 @@
...
+
+	if (infra_trims->num_entries != IFX_HPPASS_TRIMS_ENTRIES_NUM) {
+		return -EIO;
+	}
+
+	for (uint32_t i = 0; i < IFX_HPPASS_TRIMS_ENTRIES_NUM; i++) {
+		*infra_trims->entry[i].addr = infra_trims->entry[i].value;
```
Comment (teburd) [inline drivers/mfd/mfd_infineon_hppass.c:260]: So the table has both the address (a register?) to write and the value to write there? Interesting!

Perhaps a sys_write is warranted here if it is indeed a register? It would signal better to the reader (me) if so.

```c
sys_write32(infra_trims->entry[i].value, infra_trims->entry[i].addr);
```

File: drivers/mfd/mfd_infineon_hppass.c [hunk 10]
```diff
@@ -0,0 +1,1088 @@
...
+	}
+
+	for (uint8_t i = 0; i < IFX_HPPASS_SAR_CAL_LIN_TABLE_SIZE; i++) {
```
Comment (teburd) [inline drivers/mfd/mfd_infineon_hppass.c:263]: Here we are more directly writing to a register from a table in memory presumably.

Interesting that for trims we do it slightly different.

File: drivers/mfd/mfd_infineon_hppass.c [hunk 11]
```diff
@@ -0,0 +1,1088 @@
...
+ * Pulse the selected FW trigger lines for one cycle.  Only takes effect
+ * when the corresponding trigger input is configured CY_HPPASS_TR_FW_PULSE.
+ * Safe while the AC is running.
+ */
+void ifx_hppass_ac_set_fw_trigger_pulse(const struct device *dev, uint8_t mask)
+{
+	const struct ifx_hppass_mfd_config *config = dev->config;
+
+	sys_write32(mask, config->base + IFX_HPPASS_INFRA_FW_TR_PULSE);
+} /* ifx_hppass_ac_set_fw_trigger_pulse() */
```
Comment (teburd) [inline drivers/mfd/mfd_infineon_hppass.c:606]: The suffixed comments indicating the end curly for each function aren't typical Zephyr styling but I can't say I dislike it

**Review by teburd — DISMISSED** [review]
Looks really nice, and I appreciate all the doc comments around this as I mostly can infer what this is doing without even needing the reference manual which is fantastic.

Some minor comments. A great follow on to this and the autanalog from @ifx-rmcs would be *samples* showcasing how we can use these great IP blocks, whether that is for a tight faked out control loop or low power analog mic wake ups. These can be board specific samples given the hardware is uniquely ours.

File: drivers/mfd/Kconfig.infineon_hppass (hunk 2 above)
Comment (jsbatch) [reply drivers/mfd/Kconfig.infineon_hppass]: Oh right, I guess I need to update my templates.  We keep going back and forth on copyright it seems...

File: drivers/mfd/mfd_infineon_hppass.c (hunk 11 above)
Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass.c:606]: That's styling I picked up early in my career and it's stuck with me.  I find it helps me keep a better bearing of where I am in a file, particularly around longer functions.

File: drivers/mfd/mfd_infineon_hppass.c (hunk 8 above)
Comment (jsbatch) [reply drivers/mfd/mfd_infineon_hppass.c:253]: The trim table and calibration table code as written here is largely borrowed from the PDL code.  Unfortunately the TRM coverage on the calibration and trim settings is pretty sparse and largely redirects users to the PDL code.

File: drivers/mfd/Kconfig.infineon_hppass (hunk 2 above)
Comment (jsbatch) [reply drivers/mfd/Kconfig.infineon_hppass]: Updated the copyright headers.

**Review by ifx-rmcs — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
It would be nice with an in-depth review from someone outside of Infinion on this.

**Comment by teburd** [comment]
> It would be nice with an in-depth review from someone outside of Infinion on this.

No disagreement from me, but there is no ADC maintainer and very few collaborators it seems.

To be fair too @JarmouniA had approved as well previously.
